File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.py
def solution(board, moves):
    
    r_board = [[] for _ in range(len(board))] # 각 열(Column)별의 인형 리스트
    
    reversed_board = board[::-1]
    
    for b in reversed_board:
        for i in range(len(board)):
            if b[i] != 0:
                r_board[i].append(b[i])
            
    basket = []
    answer = 0
    
    for m in moves:
        col_idx = m - 1 # moves는 1부터 시작하므로 인덱스 조정
        
        # 해당 열에 인형이 있다면 뽑기
        if r_board[col_idx]:
            doll = r_board[col_idx].pop() # 가장 위에 있는 인형 꺼내기
            
            # 바구니의 마지막 인형과 같은지 확인
            if basket and basket[-1] == doll:
                basket.pop() # 같으면 제거
                answer += 2  # 인형 2개가 사라짐
            else:
                basket.append(doll) # 다르면 바구니에 추가
    
    return answer

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int n = board.length;
        
        // 각 열(Column)별 인형을 담을 Stack 리스트 생성
        List<Stack<Integer>> rBoard = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            rBoard.add(new Stack<>());
        }

        // 바닥부터 위로 올라오며 인형을 Stack에 쌓기
        for (int i = n - 1; i >= 0; i--) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] != 0) {
                    rBoard.get(j).push(board[i][j]);
                }
            }
        }

        // 인형 뽑기
        Stack<Integer> basket = new Stack<>();
        int answer = 0;

        for (int move : moves) {
            int colIdx = move - 1;
            Stack<Integer> column = rBoard.get(colIdx);

            // 해당 열에 인형이 있다면 꺼내기
            if (!column.isEmpty()) {
                int doll = column.pop();

                // 바구니가 비어있지 않고, 맨 위 인형과 방금 뽑은 인형이 같다면
                if (!basket.isEmpty() && basket.peek() == doll) {
                    basket.pop();
                    answer += 2; // 인형 2개 제거
                } else {
                    basket.push(doll);
                }
            }
        }

        return answer;
    }
}